tokio-util/src/io/sync_read_stream.rs

```diff
@@ -0,0 +1,104 @@
+use futures_core::stream::Stream;
+use std::io;
+use std::io::Read;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::task::Context;
+use std::task::Poll;
+use tokio::runtime::Handle;
+use tokio::sync::mpsc::channel;
+use tokio::sync::mpsc::Receiver;
+use tokio::sync::mpsc::Sender;
+
+const DEFAULT_BUFFER_SIZE: usize = 1024 * 16;
+
+/// Use a [`SyncReadStream`] to asynchronously read from a
+/// synchronous API.
+#[derive(Debug)]
+pub struct SyncReadStream<R: Read + Send> {
+    reader: Arc<Mutex<R>>,
+    incoming_data: Receiver<Result<bytes::Bytes, io::Error>>,
+    send_data: Sender<Result<bytes::Bytes, io::Error>>,
+    rt: Handle,
+    buffer_size: usize,
+}
+
+impl<R: Read + Send> SyncReadStream<R> {
+    /// This must be called from within a Tokio runtime context, or else it will panic.
+    #[track_caller]
+    pub fn new(rt: Handle, reader: R, buffer_size: usize) -> Self {
+        let (send_data, incoming_data) = channel(buffer_size);
+
+        Self {
+            rt,
+            buffer_size,
+            reader: Arc::new(reader.into()),
+            incoming_data,
+            send_data,
+        }
+    }
```

Comment:
There's actually a much simpler to write this, without using the expensive `spawn_blocking` in `poll_next`:

```suggestion
#[derive(Debug)]
pub struct SyncReadStream {
    rx: Receiver<Result<bytes::Bytes, io::Error>>,
    join_handle: Option<JoinHandle<()>>,
}

impl SyncReadStream {
    /// This must be called from within a Tokio runtime context, or else it will panic.
    #[track_caller]
    pub fn new<R: Read + Send + 'static>(rt: Handle, reader: R, buffer_size: NonZeroUsize) -> Self {
        let (rx, tx) = channel(1);
        let tokio_handle = handle.clone();

        Self {
            rx,
            join_handle: Some(rt.spawn_blocking(move || {
                Self::do_read_task(tx, &mut reader, buffer_size, tokio_handle);
            })),
        }
    }

    fn do_read_task(tx: crate::sync::mpsc::Sender, reader: &mut Read, buffer_size: NonZeroUsize, rt: Handle) {
        let mut buffer = bytes::BytesMut::with_capacity(buffer_size.get());

        loop {
            // TODO: Add Sender::blocking_reserve so that we don't have
            // to use Handle::block_on here
            if let Ok(permit) = handle.block_on(tx.reserve()) {
                // buffer.capacity() must not be 0, otherwise
                // we might return eof while the reader hasn't
                // reached eof yet.
                debug_assert_ne!(buffer.capacity(), 0);
                debug_assert!(buffer.is_empty());

                let res = std::io::copy(&mut reader, &mut bytes::BufMut::writer(&mut buffer));

                if let Ok(&0) = &res {
                    // eof
                    break;
                }
                
                let is_err = res.is_err();

                if !is_err {
                    // reserve for next read.
                    // Do this before split and freeze to ensure underlying
                    // allocation get reused (BytesMut can only reuse
                    // allocation if it is the only one having access to it).
                    buffer.reserve(buffer_size.get());
               }

                permit.send(res.map(|_| buffer.split().freeze()));

                if is_err {
                    break;
                }
            } else {
                // Rx is closed or dropped
                break;
            }
        } // end of loop
    }
```